File: LeetCode/1896. Minimum Cost to Change the Final Value of Expression/solution.py
class Solution:
    def minOperationsToFlip(self, expression: str) -> int:

        st = [[]]
        ops = []
        expression = '('+expression+')'
        n = len(expression)
        for i in range(n)[::-1]:
            ch = expression[i]
            # print(ch,st,ops)
            if ch == ')':
                st.append([])
            elif ch == '(':
                while len(st[-1]) >= 2:
                    lc, lv = st[-1].pop()
                    rc, rv = st[-1].pop()
                    op = ops.pop()
                    if op == '&':
                        if (lv ^ rv) == 0:
                            change = min(lc,rc)+(1-lv)
                        else:
                            change = 1
                        value = (lv&rv)
                    else:
                        if (lv ^ rv) == 0:
                            change = min(lc,rc)+lv
                        else:
                            change = 1
                        value = (lv|rv)
                    st[-1].append((change,value))
                tmp = st.pop()[-1]
                st[-1].append(tmp)

                
            elif ch in ['0','1']:
                st[-1].append((1,int(ch)))
            else:
                ops.append(ch)
        return st[-1][0][0]

File: LeetCode/1896. Minimum Cost to Change the Final Value of Expression/solution.cpp
class Solution {
public:
    int minOperationsToFlip(string expression) {
        int cur = 0;
        vector<int> depth, val, dp;
        vector<char> ops;
        for (auto ch : expression){
            if (ch == '('){
                cur++;
            }
            else if (ch == ')'){
                cur--;
                depth.back()--;
            }
            else if (ch == '|' || ch == '&'){
                ops.push_back(ch);
            }
            else{
                val.push_back(ch-'0');
                dp.push_back(1);
                depth.push_back(cur);
            }
            int n = depth.size();
            if (n >= 2 && depth[n-1] == depth[n-2]){
                depth.pop_back();
                char op = ops.back(); ops.pop_back();
                int a = val.back(); val.pop_back();
                int b = val.back(); val.pop_back();
                int x = dp.back(); dp.pop_back();
                int y = dp.back(); dp.pop_back();
                if (a^b){
                    dp.push_back(1);
                }
                else{
                    dp.push_back(min(x,y)+((a==1)^(op=='&')));
                }
                if (op == '&'){
                    val.push_back(a&b);
                }
                else{
                    val.push_back(a|b);
                }
            }
        }
        return dp.back();
    }
};